core/src/wsv.rs

```diff
@@ -60,8 +65,28 @@ pub struct World {
     pub(crate) permission_token_definitions: crate::PermissionTokenDefinitionsMap,
     /// Triggers
     pub(crate) triggers: TriggerSet,
-    /// Chain of *runtime* validators
-    pub(crate) validators: crate::validator::Chain,
+    /// Runtime Permission Validator
+    pub(crate) validator: RwLock<Option<Validator>>,
+    /// New version of Validator, which will replace `validator` on the next
+    /// [`validator_view()`}(WorldStateView::validator_view call.
+    pub(crate) upgraded_validator: RwLock<Option<Validator>>,
```

Comment:
this is new

Reply:
Yes, that fixes the deadlock bug I found, then I introduced a test for `Upgrade` instruction. The problem was that validator tried to replace it-self because we execute instructions inside validator.

Probably an idea with `AtomicPtr` suggested by @appetrosyan will help here. It's described in #3390

Reply:
makes sense. Atomic pointer should be the proper way